src/enzyme_ad/jax/Passes/MarkFunctionMemoryEffectsPass.cpp

```diff
@@ -237,26 +411,34 @@ struct MarkFunctionMemoryEffectsPass
       if (!funcOp)
         continue;
 
-      SmallVector<Attribute> effectsAttrs;
-      for (int i = 0; i < effectsSet.size(); i++) {
-        if (effectsSet[i]) {
-          if (i == 0) {
-            effectsAttrs.push_back(builder.getStringAttr("read"));
-          } else if (i == 1) {
-            effectsAttrs.push_back(builder.getStringAttr("write"));
-          } else if (i == 2) {
-            effectsAttrs.push_back(builder.getStringAttr("allocate"));
-          } else if (i == 3) {
-            effectsAttrs.push_back(builder.getStringAttr("free"));
-          } else {
-            assert(false && "unknown memory effect");
-          }
+      auto funcEffectInfo = getEffectInfo(builder, effectsSet);
+      funcOp->setAttr("enzymexla.memory_effects",
+                      funcEffectInfo.enzymexlaEffects);
+
+      auto &argEffects = funcArgEffects[symbol];
+      for (unsigned i = 0; i < funcOp.getNumArguments(); i++) {
+        auto argEffectInfo = getEffectInfo(builder, argEffects[i]);
+        funcOp.setArgAttr(i, "enzymexla.memory_effects",
```

Comment:
potentially a follow up pr, but nocapture is likely to be immensely useful too [aka, for a pointer/memref input, it isn't stored or returned]